File: app/channels/conversation_channel.rb
# frozen_string_literal: true

class ConversationChannel < ApplicationCable::Channel
  def subscribed
    stream_for Conversation.find(params[:conversation_id])
  end

  def unsubscribed
    PresenceService.mark_offline(current_user.id)
  end

  def typing(_data)
    ConversationChannel.broadcast_to(
      Conversation.find(params[:conversation_id]),
      { typing: { user_id: current_user.id, username: current_user.username } }
    )
  end
end

File: app/javascript/channels/conversation_channel.js
import consumer from "./consumer"

export default function createConversationChannel(conversationId) {
  let typingTimeout = null;
  const currentUserId = document.querySelector('meta[name="current-user-id"]')?.content;

  return consumer.subscriptions.create(
    { channel: "ConversationChannel", conversation_id: conversationId },
    {
      received(data) {
        if (data.typing) {
          if (String(data.typing.user_id) === String(currentUserId)) return;

          const indicator = document.getElementById("typing-indicator");
          if (indicator) {
            indicator.innerText = `${data.typing.username} is typing...`;
            if (typingTimeout) clearTimeout(typingTimeout);
            typingTimeout = setTimeout(() => {
              const input = document.querySelector('[data-typing-target="input"]');
              if (!input || input.value.trim() === "") {
                indicator.innerText = "";
              }
            }, 2000);
          }
        }
      }
    }
  );
}
